File: latest/week2/workshop4_m2/W04_M02_demo.R
# ============================================================
# Workshop 4 — Module 2: Data Wrangling for Visualization (R)
# R Demonstration Script — UNYT Tirana
# ============================================================
library(tidyverse); library(patchwork); dir.create("output",showWarnings=FALSE)

# ── FULL CLEANING PIPELINE ──────────────────────────────
apps_clean <- read_csv("googleplaystore.csv") |>
  distinct(App, .keep_all = TRUE) |>
  mutate(
    Reviews  = as.numeric(Reviews),
    Installs = str_remove_all(Installs, "[+,]") |> as.numeric(),
    Size     = case_when(
      str_detect(Size, "M") ~ str_remove(Size, "M") |> as.numeric(),
      str_detect(Size, "k") ~ str_remove(Size, "k") |> as.numeric() / 1024,
      TRUE ~ NA_real_),
    Price    = str_remove(Price, "\\$") |> as.numeric(),
    date     = mdy(`Last Updated`),
    year     = year(date)
  ) |>
  filter(!is.na(Rating), Type %in% c("Free", "Paid")) |>
  mutate(Category = fct_lump_n(Category, 8) |> fct_infreq())

# ── 1. PIPE-TO-PLOT: summarise → bar ────────────────────
p1 <- apps_clean |>
  group_by(Category) |>
  summarise(mean_r=mean(Rating), n=n(), .groups="drop") |>
  slice_max(n, n=8) |>
  mutate(Category=fct_reorder(Category, mean_r), hl=mean_r==max(mean_r)) |>
  ggplot(aes(x=Category, y=mean_r, fill=hl)) +
  geom_col(width=0.5, show.legend=FALSE) +
  geom_text(aes(label=round(mean_r,2)), hjust=-0.2, size=3, fontface="bold") +
  scale_fill_manual(values=c("TRUE"="#1565C0","FALSE"="#BBBBBB")) +
  coord_flip() + theme_minimal() +
  labs(title="(a) Mean Rating by Top Categories", x=NULL, y="Mean Rating")

# ── 2. PIPE-TO-PLOT: filter → scatter ───────────────────
p2 <- apps_clean |>
  filter(Reviews > 0) |>
  ggplot(aes(x=Reviews, y=Rating, color=Type)) +
  geom_point(alpha=0.2, size=0.5) +
  scale_x_log10(labels=scales::comma) +
  scale_color_manual(values=c(Free="#1565C0", Paid="#E53935")) +
  theme_minimal() + theme(legend.position="bottom") +
  labs(title="(b) Reviews vs Rating (pipe → scatter)", x="Reviews (log)", y="Rating")

# ── 3. PIVOT_LONGER DEMO ────────────────────────────────
# Simulated wide data
df_wide <- tibble(
  region = c("Tirana","Durres","Vlore","Elbasan"),
  Q1 = c(120,90,65,78), Q2 = c(135,95,72,82),
  Q3 = c(128,88,70,75), Q4 = c(150,110,80,95))

p3 <- df_wide |>
  pivot_longer(cols=Q1:Q4, names_to="quarter", values_to="revenue") |>
  ggplot(aes(x=quarter, y=revenue, color=region, group=region)) +
  geom_line(linewidth=1) + geom_point(size=2) +
  scale_color_brewer(palette="Set2") +
  theme_minimal() + labs(title="(c) After pivot_longer: Line Chart Ready")

# ── 4. GROUP_BY + SUMMARISE → POINTRANGE ────────────────
p4 <- apps_clean |>
  filter(Category != "Other") |>
  group_by(Category) |>
  summarise(mean_r=mean(Rating), se=sd(Rating)/sqrt(n()), .groups="drop") |>
  mutate(Category=fct_reorder(Category, mean_r)) |>
  ggplot(aes(x=Category, y=mean_r)) +
  geom_pointrange(aes(ymin=mean_r-1.96*se, ymax=mean_r+1.96*se), color="#1565C0", linewidth=0.8) +
  coord_flip() + theme_minimal() +
  labs(title="(d) group_by → summarise → pointrange", x=NULL, y="Mean Rating ± 95% CI")

ggsave("output/wrangling_panels.png", (p1|p2)/(p3|p4), width=12, height=9, dpi=300)

# ── 5. CLEANING BEFORE/AFTER ────────────────────────────
cat("── Before cleaning ──\n")
cat("Rows:", nrow(read_csv("googleplaystore.csv", show_col_types=FALSE)), "\n")
cat("── After cleaning ──\n")
cat("Rows:", nrow(apps_clean), "\n")
cat("Columns:", ncol(apps_clean), "\n")
cat("NA in Rating:", sum(is.na(apps_clean$Rating)), "\n")

cat("\n── All W04-M02 R plots saved ──\n")

File: latest/week2/workshop4_m2/W04_M02_demo.py
"""W04-M02: Data Wrangling for Visualization (R focus) — Python equivalent — UNYT"""
import numpy as np, pandas as pd, matplotlib.pyplot as plt, os
os.makedirs("output",exist_ok=True); np.random.seed(42)

# Full cleaning pipeline (pandas equivalent of the R pipe chain)
apps = pd.read_csv("googleplaystore.csv")
apps_clean = (apps
    .drop_duplicates(subset="App", keep="first")
    .assign(
        Reviews=lambda d: pd.to_numeric(d["Reviews"], errors="coerce"),
        Installs=lambda d: d["Installs"].str.replace(r"[+,]","",regex=True).pipe(pd.to_numeric, errors="coerce"),
        Size_MB=lambda d: d["Size"].str.replace("M","").pipe(pd.to_numeric, errors="coerce"),
        Price=lambda d: d["Price"].str.replace("$","").pipe(pd.to_numeric, errors="coerce"),
        date=lambda d: pd.to_datetime(d["Last Updated"], format="%B %d, %Y", errors="coerce"))
    .query("Rating == Rating and Type in ['Free','Paid']")  # drop NA Rating
    .copy())
print(f"Clean: {apps_clean.shape[0]} rows, {apps_clean.shape[1]} cols")

# 1. PIPE-TO-PLOT: groupby → bar
summary = (apps_clean.groupby("Category")
    .agg(mean_r=("Rating","mean"), n=("Rating","count"))
    .nlargest(8,"n").sort_values("mean_r"))
fig,ax=plt.subplots(figsize=(6,5))
colors=["#BBBBBB"]*8; colors[-1]="#1565C0"
ax.barh(summary.index, summary["mean_r"], color=colors, height=0.5)
for i,(idx,row) in enumerate(summary.iterrows()):
    ax.text(row["mean_r"]+0.005,i,f"{row['mean_r']:.2f}",va="center",fontsize=7,fontweight="bold",
        color="#1565C0" if i==7 else "#888")
ax.set_xlim(3.8,4.4); ax.set_title("df.groupby().agg().nlargest().plot()",fontweight="bold")
for sp in ["top","right","left"]: ax.spines[sp].set_visible(False)
ax.tick_params(left=False,labelsize=7); ax.grid(axis="x",alpha=0.1)
plt.tight_layout(); plt.savefig("output/pipe_to_plot_py.png",dpi=300); plt.close()

# 2. PIVOT (melt) → line chart
df_wide = pd.DataFrame({"region":["Tirana","Durres","Vlore","Elbasan"],
    "Q1":[120,90,65,78],"Q2":[135,95,72,82],"Q3":[128,88,70,75],"Q4":[150,110,80,95]})
df_long = df_wide.melt(id_vars="region", var_name="quarter", value_name="revenue")
fig,ax=plt.subplots(figsize=(6,4))
pal={"Tirana":"#1565C0","Durres":"#E53935","Vlore":"#2E7D32","Elbasan":"#E65100"}
for r,c in pal.items():
    sub=df_long.query("region==@r")
    ax.plot(sub["quarter"],sub["revenue"],"-o",color=c,lw=1.5,markersize=5,label=r)
ax.legend(fontsize=7); ax.set_title("After pd.melt(): Line Chart Ready",fontweight="bold")
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
plt.tight_layout(); plt.savefig("output/pivot_line_py.png",dpi=300); plt.close()

# 3. GROUPBY → POINTRANGE
cats=apps_clean["Category"].value_counts().head(8).index
sub=apps_clean.query("Category in @cats")
stats=(sub.groupby("Category")["Rating"]
    .agg(["mean","std","count"])
    .assign(se=lambda d: d["std"]/np.sqrt(d["count"]),
            ci95=lambda d: 1.96*d["std"]/np.sqrt(d["count"]))
    .sort_values("mean"))
fig,ax=plt.subplots(figsize=(6,4.5))
ax.errorbar(stats["mean"],range(len(stats)),xerr=stats["ci95"],fmt="o",color="#1565C0",
    markersize=6,capsize=4,lw=1.5)
ax.set_yticks(range(len(stats))); ax.set_yticklabels(stats.index,fontsize=7)
ax.set_title("groupby → agg → errorbar (Mean ± 95% CI)",fontweight="bold")
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
plt.tight_layout(); plt.savefig("output/pointrange_py.png",dpi=300); plt.close()

print("All W04-M02 Python plots saved")
